### ISSUE TITLE
Benchmarking

### ISSUE THREAD

**Issue by superkhau** [strongloop/loopback#1291]
- [Wiki](https://github.com/strongloop/loopback/wiki/Benchmarks)
- [Google Doc](https://docs.google.com/document/d/11H2huId0W6g_PLA2ykCHVfDXV1usb6NiodQvXnHAKZQ/edit#)

---

I would like to open a discussion on our strategy for applying benchmarks across various LoopBack repos as discussed during our meeting @raymondfeng and @ritch.

/cc @bajtos

**Comment by superkhau** [comment]
Actionable items:

- [ ] Create Wiki to gather discussion ideas
- [ ] Spike on possible tooling to automate benchmarking (something like `mocha` for benchmarks)
- [ ] Draft of strategy discussed during meeting with @raymondfeng and @ritch
  - [ ] Discuss which repos to cover
  - [ ] Discuss the ordering/priority of repos to start implementation
  - [ ] Discuss what level of benchmarking to start with
- [ ] Pass v1 of the wiki around the engineering team for reviewing
- [ ] Create next actionable items list

**Comment by superkhau** [comment]
@bajtos @raymondfeng @ritch Please edit the [wiki](https://github.com/strongloop/loopback/wiki/Benchmarks) and add your input. I will clean up the draft again once you guys add your parts.

**Comment by raymondfeng** [comment]
@superkhau Do you want to check out https://github.com/bestiejs/benchmark.js?

**Comment by superkhau** [comment]
Sure, I'll give it a shot tomorrow. This lib seems more geared towards performance testing.

**Comment by superkhau** [comment]
- [ ] Spike on https://github.com/bestiejs/benchmark.js

**Comment by bajtos** [comment]
I'll comment here, wiki is really bad for discussions (Google Docs works better for that IMO).

I would personally start with a full end-to-end benchmark covering loopback, strong-remoting, loopback-datasource-juggler and and a single connector. The connector under test should be configurable via the environment or command-line arguments, no code changes needed.

This should give us a good overview of which parts work well and which needs most improvement.

For the beginning, I would focus on the following scenarios:
 1. Detect memory leaks
 2. Benchmark requests per second, measure how much time is needed by individual parts: strong-remoting, loopback+juggler, connector

Example usage:

```js
$ gulp memtest
(...runnning...)
No memory leaks detected

$ gulp benchmark mysql
(...running...)
Requests per second:
  median: 1700
  stddev: 300

Times per request (wall time in microseconds)
             total      strong-remoting      loopback     connector
median        588                  200            80            308       
90p           812                  300           200            312
99p           932                  350           240            342
````

The made-up numbers above show that we spend most time talking to the database, so that's probably the first thing to check when looking for places to optimize. The second biggest contributor is strong-remoting, especially when the load gets higher, so that would be the second place to investigate.

**Comment by ritch** [comment]
> Google Docs works better for that IMO

+9999

**Comment by ritch** [comment]
Overall I would love to have the tooling suggested by @bajtos. But I'm afraid that it will lead to scenarios like this fairly often:

 1. foo submits a PR to strong-remoting that decreases the benchmark numbers
 2. the benchmarks are not run (since its takes a bit of time to do and it isnt automatic)... so impact in performance is completely missed
 3. the PR is released
 4. someone notices the performance decrease by running the lb benchmark with the new version of remoting
 5. there is a discussion on how to revert the new feature added by foobar
 6. we can't since it would be a breaking change
 7. :(

**Comment by ritch** [comment]
Instead we should have automated benchmarks run per module. We should be able to see if there is a decrease on a per PR basis. Otherwise I'm not sure its worth the effort.

This shouldn't be that hard to accomplish IMO.

**Comment by superkhau** [comment]
Transferring wiki data into a Google doc as per your suggestions from above.

+9999 for automation (at least run it periodically if every change takes too long to benchmark)

**Comment by superkhau** [comment]
@bajtos @ritch [Google Doc](https://docs.google.com/document/d/11H2huId0W6g_PLA2ykCHVfDXV1usb6NiodQvXnHAKZQ/edit#) created. We can comment directly there now.

/cc @raymondfeng

**Comment by superkhau** [comment]
@bajtos @raymondfeng @ritch Ryan will need to spend some time setting up CI in a few different ways for us to run benchmarks. Some items of interest from our discussion include:

- We should come up with a standard to trigger benchmark builds (`npm run benchmarks`)
- Leak-detection should be a part of the testing infrastructure
  - Long running tests may back up regular CI, so it was suggested we run these as a separate Jenkins job
  - Maybe `npm run leak-detection` or something along those lines (this only runs leak-detection tests)
- Jenkins supports archiving of text-based files, meaning we can view benchmark results as long as we output to a standard file (`benchmark.json`,`benchmark.log`, etc)

For now, I will write:

1. Leak-detection test for `find` and `create` using the mongo connector
2. Performance test for `find` and `create`using the mongo connector

/cc @rmg @chandadharap

**Comment by rmg** [comment]
Trivial leaks shouldn't require long test runs and non-trivial leaks probably can't be reliably detected by CI, so I'm not sure it is worth the effort trying to tackle the problem of long-running leak tests.

**Comment by superkhau** [comment]
@rmg Turns out the tests don't take too long to run at the moment (it can take longer if we increase the number of iterations). Maybe we should just add the tests to regular CI instead? Anyways, we'll see what @raymondfeng says after reviewing the spike code.

**Comment by superkhau** [comment]
MongoDB spike for benchmarks and leak detection

Connect to strongloop/loopback-connector-mongodb#126
Connect to strongloop/loopback-connector-mongodb#124

**Comment by rmg** [comment]
@superkhau CI runs `npm test`, so they should be added to the `mocha-and-karma` grunt task if CI is to run them.

**Comment by superkhau** [comment]
@rmg Sounds good, I'll wait to see what the rest of the team says before making the changes to grunt.

**Comment by bajtos** [comment]
> Turns out the tests don't take too long to run at the moment (it can take longer if we increase the number of iterations). Maybe we should just add the tests to regular CI instead?

If the leak detection runs in the order of seconds (~10s max), then I am all for including in `npm test` runs.

**Comment by superkhau** [comment]
>If the leak detection runs in the order of seconds (~10s max)...

For the MongoDB tests we have so far, each test takes about 250ms for 100 iterations and 2000ms for 1000 iterations. This will obviously increase as we add more functions to test, so your max of 10s will allow us to test 40 functions at 100 iterations or 5 functions at 1000 iterations.

**Comment by bajtos** [comment]
> For the MongoDB tests we have so far, each test takes about 250ms for 100 iterations and 2000ms for 1000 iterations. This will obviously increase as we add more functions to test, so your max of 10s will allow us to test 40 functions at 100 iterations or 5 functions at 1000 iterations.

That's sounds very reasonable to me, providing that 100 iterations are enough to discover a memory leak (they should be).

**Comment by superkhau** [comment]
Benchmark prototypes merged. Now that we have infrastructure in place, we need to decide which repos to cover over time.

- [ ] Benchmarks
  - [ ] Which functions will we benchmark
  - [ ] Which repos will we cover
- [ ] Leak detection
  - [ ] Which functions to cover
  - [ ] Which repos will we cover

For example, we cover `find`, `find with options`, and `create` at the moment for benchmarks in the mongodb connector and leak detection covers only `create` and `find`. Should we repeat for all the other connectors, etc.

@bajtos @raymondfeng @ritch ^

**Comment by chandadharap** [comment]
@raymond please respond to the above question.

**Comment by raymondfeng** [comment]
1. It would be ideal to cover all major modules, such as loopback, juggler, strong-remoting, and connectors
2. We should have integration scenarios for benchmarking and lead detection

**Comment by bajtos** [comment]
From my POV, I'd like to see a benchmark that will tell us which part of LoopBack is the slowest one (see https://github.com/strongloop/loopback/issues/1291#issuecomment-91280070). For example, start with the method "find with options" and gather timing data split across module boundaries: express, strong-remoting, loopback, loopback-datasource-juggler, connector, SQL backend.

**Comment by chandadharap** [comment]
Have to move back. Commented and feedback done, Simon on a longish PTO.

**Comment by superkhau** [comment]
@raymondfeng @bajtos There are many stories involved here. I see two separate issues:

Leak detection
- Create a new project named `loopback-leak-detection
- Individual package that can be run on demand
- Detect leaks for new functions added over time on an as-needed basis
- Integration style tests (ie. maybe a dir for each functional area we want to check for leaks)

The reason for this is to do leak detection on every function would be an exponential problem as each leak detection test needs to be written in a a fashion similar to unit testing. The running the tests longer also do a better job of detecting leaks. So my suggestion for the entire solution is:

1. New project to consolidate tests for all loopback projects (this project will do leak detection for various parts of LB, strong-remoting, connectors, etc)
2. Pushing this project to GitHub should trigger the CI to run tests (which can take hours depending on the number of iterations we decide on)

The benefits of doing this way is that we can just add tests as need and the CI can run tests as it normally does without major infrastructure changes. This CI server should be **separate** from the regular CI server because these long running jobs would block other jobs from executing on the normal CI server.

**Comment by superkhau** [comment]
>From my POV, I'd like to see a benchmark that will tell us which part of LoopBack is the slowest one (see #1291 (comment)). For example, start with the method "find with options" and gather timing data split across module boundaries: express, strong-remoting, loopback, loopback-datasource-juggler, connector, SQL backend.

I agree with this approach as we're not just benchmarking randomly. Maybe start with benchmarking each connector vs it's alternative (ie. `loopback-connector-mongodb` vs `native` vs `mongoose`). Do the same for all 5 connectors. Then we can perform the same comparisons with juggler + connectors. Then once we have this, we can add in strong remoting, etc. This way, we can split this giant task into smaller stories and flush it out over time.

Story 1 - mongo connector vs mongoose - 2 points
Story 2 - mongo connector vs mongo native - 2 points
Story 3 - juggler + mongo connector vs mongoose - 2 points
Story 4 - juggler + mongo connector vs mongo native - 2 points
Story 5 - remoting + juggler + mongo connector - 2 points
Story 6 - express + mongoose - 2 points
Sotry 7 - express + mongo native - 2 points

And so forth. We still have to decide where to persist the results to make it easy to compare the results.

**Comment by superkhau** [comment]
Spike at https://github.com/strongloop/loopback-benchmark-mongodb

**Comment by superkhau** [comment]
@bajtos @raymondfeng Please review https://github.com/strongloop/loopback-benchmark-mongodb/blob/master/results.md and debate whether its worth continuing. I only completed the time-split data for express-mongodb, will add express-mongoose and loopback-loopback-connector-mongodb if you guys think its worth continuing (so you can compare full request split-times).

**Comment by bajtos** [comment]
If I am reading the benchmark results correctly, then our MongoDB connector is performing roughly as well as mongoose/mongodb and there isn't much space for optimization at our side. Is my reasoning correct?

> if you guys think its worth continuing (so you can compare full request split-times).

I think it would be good to create a similar benchmark for one of our SQL connectors too. Then I would personally like to see how much overhead is added by juggler and strong-remoting. So yes, I think it makes sens to continue.

**Comment by superkhau** [comment]
@bajtos Yes, you are reasoning correctly from the subset of data provided. I wanted to make sure this spike is providing the data you want to see before diving too deeply.

The plan is to do the same for the the other connectors, but I would like to finish up the MongoDB spike first to flush out any issues or questions that may need addressing. Adding the last two scenarios would make this a fairer comparision because `express-mongoose` is a common strategy and of course `loopback-loopback-connector-mongodb` would be a better real world baseline.

After this, we need to figure out a strategy for versioning. The way I set it up, I'm thinking we can use git tags for updates, so we can jump back and forth between updates and keep a history based on major comparisons we've decided on over time.
